Nightly integration tests failing - GKE & Kubernetes v1.21.0 (test-previous-kubernetes-gke)

Integration tests in the nightly workflow (test-previous-kubernetes-gke) are failing due to the following error: 
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x234ceb2]

goroutine 1491 [running]:
github.com/kong/kubernetes-ingress-controller/v2/internal/store.Store.GetIngressClassParametersV1Alpha1(***0x32a9c58, 0xc000c343d8***, ***0x32a9c58, 0xc000c343f0***, ***0x32a9c58, 0xc000c34408***, ***0x32a9c58, 0xc000c34420***, ***0x32a9c58, 0xc000c34438***, ...***, ...***)
	/home/runner/work/kubernetes-ingress-controller/kubernetes-ingress-controller/internal/store/store.go:868 +0x1b2
github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane/parser.getIngressClassParametersOrDefault(...)
	/home/runner/work/kubernetes-ingress-controller/kubernetes-ingress-controller/internal/dataplane/parser/parser.go:679
github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane/parser.(*Parser).ingressRulesFromIngressV1beta1(0xc0005e1a98)
	/home/runner/work/kubernetes-ingress-controller/kubernetes-ingress-controller/internal/dataplane/parser/translate_ingress.go:24 +0x9a
github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane/parser.(*Parser).Build(0xc0005e1a98)
	/home/runner/work/kubernetes-ingress-controller/kubernetes-ingress-controller/internal/dataplane/parser/parser.go:72 +0x36
github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane.(*KongClient).Update(0xc000c60180, ***0x329ce60, 0xc000a23440***)
	/home/runner/work/kubernetes-ingress-controller/kubernetes-ingress-controller/internal/dataplane/kong_client.go:314 +0x385
github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane.(*Synchronizer).startUpdateServer(0xc000b375c0, ***0x329ce60, 0xc000a23440***)
	/home/runner/work/kubernetes-ingress-controller/kubernetes-ingress-controller/internal/dataplane/synchronizer.go:162 +0xdd
created by github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane.(*Synchronizer).Start
	/home/runner/work/kubernetes-ingress-controller/kubernetes-ingress-controller/internal/dataplane/synchronizer.go:100 +0x14a
FAIL	github.com/kong/kubernetes-ingress-controller/v2/test/integration	435.831s
```

Sample run: https://github.com/Kong/kubernetes-ingress-controller/actions/runs/3102816618/jobs/5025633520#step:5:674

It's probably caused by dereferencing `ingressClass.Spec.Parameters.Namespace` which is optional when `ingressClass.Spec.Parameters.Scope` is not set to `Namespace`: https://github.com/Kong/kubernetes-ingress-controller/blob/94969ef3dbb82695edc787cb6c69be41de7bdd30/internal/store/store.go#L868

--- a/internal/dataplane/parser/ingressclassparemeters_test.go
+++ b/internal/dataplane/parser/ingressclassparemeters_test.go
@@ -1,0 +1,158 @@
+package parser
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	netv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/store"
+	configurationv1alpha1 "github.com/kong/kubernetes-ingress-controller/v2/pkg/apis/configuration/v1alpha1"
+)
+
+func Test_GetIngressClassParameters(t *testing.T) {
+	var (
+		scopeNamespace    = "Namespace"
+		testNamespaceName = "test-ns"
+		testIcpName       = "test-icp"
+	)
+
+	defaultIcpSpec := &configurationv1alpha1.IngressClassParametersSpec{}
+	icp := &configurationv1alpha1.IngressClassParameters{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: testNamespaceName,
+			Name:      testIcpName,
+		},
+		Spec: configurationv1alpha1.IngressClassParametersSpec{
+			EnableLegacyRegexDetection: true,
+		},
+	}
+
+	testCases := []struct {
+		name          string
+		paramRef      *netv1.IngressClassParametersReference
+		parameterSpec *configurationv1alpha1.IngressClassParametersSpec
+		hasError      bool
+	}{
+		{
+			name:          "nil-paramref",
+			parameterSpec: defaultIcpSpec,
+			hasError:      true,
+		},
+		{
+			name: "nil-apigroup",
+			paramRef: &netv1.IngressClassParametersReference{
+				Kind: "configMap",
+				Name: "some-cm",
+			},
+			parameterSpec: defaultIcpSpec,
+			hasError:      true,
+		},
+		{
+			name: "nil-scope",
+			paramRef: &netv1.IngressClassParametersReference{
+				APIGroup:  &configurationv1alpha1.GroupVersion.Group,
+				Kind:      configurationv1alpha1.IngressClassParametersKind,
+				Namespace: &testNamespaceName,
+				Name:      testIcpName,
+			},
+			parameterSpec: defaultIcpSpec,
+			hasError:      true,
+		},
+		{
+			name: "nil-namespace",
+			paramRef: &netv1.IngressClassParametersReference{
+				APIGroup: &configurationv1alpha1.GroupVersion.Group,
+				Kind:     configurationv1alpha1.IngressClassParametersKind,
+				Scope:    &scopeNamespace,
+				Name:     testIcpName,
+			},
+			parameterSpec: defaultIcpSpec,
+			hasError:      true,
+		},
+		{
+			name: "matched-parameters",
+			paramRef: &netv1.IngressClassParametersReference{
+				APIGroup:  &configurationv1alpha1.GroupVersion.Group,
+				Kind:      configurationv1alpha1.IngressClassParametersKind,
+				Scope:     &scopeNamespace,
+				Namespace: &testNamespaceName,
+				Name:      testIcpName,
+			},
+			parameterSpec: &icp.Spec,
+		},
+		{
+			name: "unmatched-kind",
+			paramRef: &netv1.IngressClassParametersReference{
+				APIGroup:  &configurationv1alpha1.GroupVersion.Group,
+				Kind:      "SomeKind",
+				Scope:     &scopeNamespace,
+				Namespace: &testNamespaceName,
+				Name:      testIcpName,
+			},
+			parameterSpec: defaultIcpSpec,
+			hasError:      true,
+		},
+		{
+			name: "unmatched-namespace",
+			paramRef: &netv1.IngressClassParametersReference{
+				APIGroup:  &configurationv1alpha1.GroupVersion.Group,
+				Kind:      configurationv1alpha1.IngressClassParametersKind,
+				Scope:     &scopeNamespace,
+				Namespace: new(string),
+				Name:      testIcpName,
+			},
+			parameterSpec: defaultIcpSpec,
+			hasError:      true,
+		},
+		{
+			name: "unmatched-name",
+			paramRef: &netv1.IngressClassParametersReference{
+				APIGroup:  &configurationv1alpha1.GroupVersion.Group,
+				Kind:      configurationv1alpha1.IngressClassParametersKind,
+				Scope:     &scopeNamespace,
+				Namespace: &testNamespaceName,
+				Name:      "another-icp",
+			},
+			parameterSpec: defaultIcpSpec,
+			hasError:      true,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			ingressClass := &netv1.IngressClass{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: testNamespaceName,
+					Name:      tc.name,
+				},
+				Spec: netv1.IngressClassSpec{
+					Parameters: tc.paramRef,
+				},
+			}
+			cacheStores, err := store.NewCacheStoresFromObjs()
+			assert.NoError(t, err)
+			err = cacheStores.Add(ingressClass)
+			assert.NoError(t, err)
+			err = cacheStores.Add(icp)
+			assert.NoError(t, err)
+			s := store.New(cacheStores, ingressClass.Name, true, true, true, logrus.New())
+			icpSpec, err := getIngressClassParametersOrDefault(s)
+			assert.Truef(t, reflect.DeepEqual(*tc.parameterSpec, icpSpec),
+				fmt.Sprintf("should get same ingress parameter spec: expected %+v, actual %+v", tc.parameterSpec, icpSpec),
+			)
+			if tc.hasError {
+				assert.NotNil(t, err)
+				assert.Truef(t, errors.As(err, &store.ErrNotFound{}), "error should be a store.ErrNotFound, actual %T", err)
+			} else {
+				assert.Nil(t, err)
+			}
+		})
+	}
+}

--- a/internal/store/store.go
+++ b/internal/store/store.go
@@ -849,7 +849,8 @@ func (s Store) GetIngressClassParametersV1Alpha1() (*kongv1alpha1.IngressClassPa
 		return nil, ErrNotFound{fmt.Sprintf("IngressClass %s doesn't reference any parameters", s.ingressClass)}
 	}
 
-	if *ingressClass.Spec.Parameters.APIGroup != kongv1alpha1.GroupVersion.Group {
+	if ingressClass.Spec.Parameters.APIGroup == nil ||
+		*ingressClass.Spec.Parameters.APIGroup != kongv1alpha1.GroupVersion.Group {
 		return nil, ErrNotFound{fmt.Sprintf(
 			"IngressClass %s should reference parameters in apiGroup:%s",
 			s.ingressClass,
@@ -863,6 +864,12 @@ func (s Store) GetIngressClassParametersV1Alpha1() (*kongv1alpha1.IngressClassPa
 			s.ingressClass,
 			kongv1alpha1.IngressClassParametersKind,
 		)}
+	}
+
+	if ingressClass.Spec.Parameters.Scope == nil || ingressClass.Spec.Parameters.Namespace == nil {
+		return nil, ErrNotFound{
+			message: fmt.Sprintf("IngressClass %s should reference namespaced parameters", ingressClass),
+		}
 	}
 
 	key := fmt.Sprintf("%v/%v", *ingressClass.Spec.Parameters.Namespace, ingressClass.Spec.Parameters.Name)

